src/openfermion/measurements/_bayesian_estimators.py

```diff
@@ -0,0 +1,805 @@
+            0.25 * cos(beta - meas * pi) * cos_matrix.dot(vectors) +\
+            0.25 * sin(beta - meas * pi) * sin_matrix.dot(vectors)
+
+    def _mlikelihood(self, a_vec):
+        # Calculates the negative likelihood of a set of amplitudes
+        # generating the observed measurements
+        return -sum([numpy.log(p_vec.dot(a_vec[:len(p_vec)]))
+                     for p_vec in self._p_vecs]) +\
+            self._init_mlikelihood(a_vec)
+
+    def _diff_mlikelihood(self, a_vec):
+        # The derivative of the above function
+        return -sum([p_vec/(p_vec.dot(a_vec[:len(p_vec)]))
+                     for p_vec in self._p_vecs]) +\
+            self._dinit_mlikelihood(a_vec)
+
+    def _single_diff(self, p_vec, a_vec):
+        # The first derivative of a single term in the
+        # likelihood function.
+        return -p_vec / numpy.dot(p_vec, a_vec[:len(p_vec)])
+
+    def _jacobian_term(self, p_vec, a_vec):
+        # The second derivative of the above function,
+        # evaluated for a single p vector - to evaluate
+        # and store
+        return numpy.dot(p_vec[:, numpy.newaxis], p_vec[numpy.newaxis, :]) /\
+            numpy.dot(p_vec, a_vec[:len(p_vec)]) ** 2
+
+    def _calculate_jacobian(self):
+        # Recalculates the Jacobian based on all previously
+        # calculated p_vecs.
+        self._jacobian = sum(
+            [self._jacobian_term(p_vec, self._amplitude_estimates)
+             for p_vec in self._p_vecs]) + self._jinit_mlikelihood()
+
+    def _init_mlikelihood(self, a_vec):
+        # A normal distributed initial guess at the amplitudes
+        return numpy.dot(
+            0.5*(a_vec-self._amplitude_guess)[numpy.newaxis, :],
+            numpy.dot(self._inv_covar_mat,
+                      (a_vec-self._amplitude_guess)[:, numpy.newaxis])).item()
+
+    def _dinit_mlikelihood(self, a_vec):
+        # The derivative of the above likelihood function
+        return numpy.dot(self._inv_covar_mat, (a_vec-self._amplitude_guess))
+
+    def _jinit_mlikelihood(self):
+        # The jacobian of the above likelihood function
+        return self._inv_covar_mat
+
+    def _make_projector(self):
+        # Calculates the projector onto the plane sum_j|a_j|^2=0
+        num_vectors = self._num_vectors
+        self._projector = numpy.identity(num_vectors) -\
+            1/num_vectors * numpy.ones([num_vectors, num_vectors])
+
+    def _make_dist(self, vector, angle):
+        """
+        converts a frequency vector into a function evaluated at theta
+        Args:
+            vector (numpy array): a fourier vector P(phi)
+            angle (float): the angle to evaluate P at
+
+        Returns:
+            (float) P(angle)
+        """
+        res = vector[0]
+        for j in range(1, self._num_freqs):
+            res += vector[2*j-1] * sin(j*angle)
+            res += vector[2*j] * cos(j*angle)
+        return res / (2*pi)
+
+
+class BayesEstimator(FourierProbabilityDist):
```

Comment:
I'm not sure about having BayesEstimator be a subclass of FourierProbabilityDist. It seems to be mixing together two quite distinct concepts: the concept of a probability distribution, and the concept of something which maintains a probability distribution internally and uses it to estimate values of things. I think it makes more sense for BayesEstimator to keep an instance of FourierProbabilityDist, which it updates as it gets more information and which it uses to obtain phase estimates. What do you think?

Reply:
tl;dr: I agree with both this and the above change, but looking through this I think there's a more broad shuffle of the contents of the two classes needed. I'll implement this now, let me know what you think.

So, to make my response more general (and respond to your above comment at the same time), I think that the interface between these two classes is a really good idea to discuss. These started off as a single class, and I split them into two at some point to try to keep track of everything, but I'm not convinced I got the split correct.

I guess there are a few different parts to this; there's the data stored just for the actual probability distribution ( _fourier_vectors, _amplitude_estimates and _amplitude_vars), there's the functions to do simple arithmetic operations on them (_make_matrices, _vector_product, _make_dist, _holevo_centers, _holevo_variances and their dependencies). Then there's the data structures that store additional data from a series of experiments (_p_vecs), and the functions that update a probability distribution given more experimental data (the update function, but also all the functions from _mlikelihood through to _make_projector).

After writing this it seems like a pretty clear distinction; everything that isn't involved directly with _fourier_vectors, _amplitude_estimates or _amplitude_vars should be chucked out of the ProbabilityDist function, and then this can be indeed treated as a separate object. In which case I might want to go about minimizing the interface between the two (but this is probably a good thing to do).

Reply:
I've implemented the changes, but I feel like the dataflow between the two classes is quite large as all the manipulation routines are now in the BayesEstimator class.